### PR TITLE
[TFLite][Testing] Add infra to write TFLite model buffers directly

### PR DESCRIPTION
**Pull request by ekalda**
This patch adds infrastructure to directly generate TFLite model buffers
by using flatc, the flatbuffers command line tool. This gives us more
freedom in creating the models for testing since we don't have to
rely on any of the converters.

* Add classes and helper functions to create the model buffer
* Add some convolution tests that test TFLite2 models in int8
  with per channel and per tensor quantization and remove the
  orphaned Keras tests

Co-authored with @NicolaLancellotti

### CODE REVIEW

**Comment by ANSHUMAN87** [comment]
> This patch adds infrastructure to directly generate TFLite model buffers
> by using flatc, the flatbuffers command line tool. This gives us more
> freedom in creating the models for testing since we don't have to
> rely on any of the converters.
> 
> * Add classes and helper functions to create the model buffer
> * Add some convolution tests that test TFLite2 models in int8
>   with per channel and per tensor quantization and remove the
>   orphaned Keras tests
> 
> Co-authored with @NicolaLancellotti

Hi @ekalda, I am just unable to see the need for such changes. As per my understanding, TFlite framework behaviour is not something we should control in TVM.
Model buffers should be created using standard Apis in TFlite. We should not use a custom one to validate our requirements which may result in failure of complete TFlite frontend Parser.

Maybe if you share what was the actual motivation for this change, we can discuss about the solution better. Thanks!

**Comment by ekalda** [comment]
> > This patch adds infrastructure to directly generate TFLite model buffers
> > by using flatc, the flatbuffers command line tool. This gives us more
> > freedom in creating the models for testing since we don't have to
> > rely on any of the converters.
> > 
> > * Add classes and helper functions to create the model buffer
> > * Add some convolution tests that test TFLite2 models in int8
> >   with per channel and per tensor quantization and remove the
> >   orphaned Keras tests
> > 
> > Co-authored with @NicolaLancellotti
> 
> Hi @ekalda, I am just unable to see the need for such changes. As per my understanding, TFlite framework behaviour is not something we should control in TVM.
> Model buffers should be created using standard Apis in TFlite. We should not use a custom one to validate our requirements which may result in failure of complete TFlite frontend Parser.
> 
> Maybe if you share what was the actual motivation for this change, we can discuss about the solution better. Thanks!

Hi @ANSHUMAN87, see that RFC for some more motivation - https://discuss.tvm.apache.org/t/rfc-tflite-frontend-create-models-for-frontend-testing-by-directly-writing-tflite-buffers/9811

The gist is that the current converters that convert into TFLite are just not flexible enough when it comes to creating the one operator models with various properties (e.g. different fused activations). We have found that writing buffers directly is the most convenient, fast and debuggable way for consistently generating various one operator models with desired properties. 

As of whether the models created like that are valid TFLite models - since we use the TFLite schema to create the buffers, all the models created this way are valid TFLite models and if TVM frontend fails to parse them, that indicates problem with the TVM's TFLite's frontend parser. 

Also tagging @mbaret @FrozenGene @manupa-arm @anijain2305 @leandron

**Comment by ANSHUMAN87** [comment]
> > > This patch adds infrastructure to directly generate TFLite model buffers
> > > by using flatc, the flatbuffers command line tool. This gives us more
> > > freedom in creating the models for testing since we don't have to
> > > rely on any of the converters.
> > > 
> > > * Add classes and helper functions to create the model buffer
> > > * Add some convolution tests that test TFLite2 models in int8
> > >   with per channel and per tensor quantization and remove the
> > >   orphaned Keras tests
> > > 
> > > Co-authored with @NicolaLancellotti
> > 
> > 
> > Hi @ekalda, I am just unable to see the need for such changes. As per my understanding, TFlite framework behaviour is not something we should control in TVM.
> > Model buffers should be created using standard Apis in TFlite. We should not use a custom one to validate our requirements which may result in failure of complete TFlite frontend Parser.
> > Maybe if you share what was the actual motivation for this change, we can discuss about the solution better. Thanks!
> 
> Hi @ANSHUMAN87, see that RFC for some more motivation - https://discuss.tvm.apache.org/t/rfc-tflite-frontend-create-models-for-frontend-testing-by-directly-writing-tflite-buffers/9811
> 
> The gist is that the current converters that convert into TFLite are just not flexible enough when it comes to creating the one operator models with various properties (e.g. different fused activations). We have found that writing buffers directly is the most convenient, fast and debuggable way for consistently generating various one operator models with desired properties.
> 
> As of whether the models created like that are valid TFLite models - since we use the TFLite schema to create the buffers, all the models created this way are valid TFLite models and if TVM frontend fails to parse them, that indicates problem with the TVM's TFLite's frontend parser.
> 
> Also tagging @mbaret @FrozenGene @manupa-arm @anijain2305 @leandron

Thanks @ekalda for detailed response.
I will go through little deeper about the point you mentioned about flexibility. 

But at high level, I am still not convinced for the approach. Please find below few fruits for thought.
1. TFlite was initially not designed to create models with first hand. Now it is coming up with model maker of its own. So the basic idea was convert existing tf and keras models to TFlite and then do certain optimization like layer fusions and post training quantization or fine tuning. And prep for inference. Now if the conversion to TFlite does not support certain features, then definitely there will not be any TFlite models with that feature.
2. Hence if you want that feature to be part of TFlite, then that has to be first part of Tensorflow project not TVM.
3. As per you mentioned, you need operator with fused activation, then again the questions arise for WHICH one? TFlite, Tensorflow or Keras. if TFlite already supports, then no point of this change. But if Tensorflow or Keras supports, but TFlite does not, then why TVM TFlite frontend needs to support?

Hope my points are clear. Please revert back in case anything. Definitely I might be missing the point which you might have observed. Please enlighten me, so that I also can be on same page. TIA!

**Comment by ANSHUMAN87** [comment]
Thanks @ekalda for such detailed response. Quite enlightened with your POV now.
I believe most of the points we both concur.
However below are some points where  my POV differs.

As you have taken example of "TFLite_PostProcess_Detection". I will consider that to state my point.
   I agree this Op is not part of Tensorflow Op mainstream and they use flatbuffer to create it. 
   But this Op is created using standard APIs from user perspective, today it is created with features x, y, h, w (just to explain) when the TF model SSD operations are present as part of it.
   Now consider later on, when this SSD support has to be upgraded from [x, y, h, w] to [x, y, w, h], which is done because there was a limitation in SSD op support. Now this limitation was visible at higher layers operation, not in the model creation. 
   So, in that case, the flatbuffer implemented "TFLite_PostProcess_Detection" in TVM will become obsolete implementation. And it will be very difficult to findout the reason, unless you are looking into the flatbuffer custom op implemenation in Tensorflow / TLite MLIR compiler project.
   Now above scenario is the failure of TFLite parser I am trying to project here, if it is validated using TVM flatbuffer implementation.  This issue can be avoided easily if we use standard APIs and the standard models which are created using those.

   Also, TFLite converters are the standard entry points into the world of TFLite and they are stable w.r.t. their own domain of versions(for some operators the dev takes time and it gets stretched through versions). But there is no TFLite standard(highlight) model exists which has bypassed it. 
  

NOTE: I do clearly understand that testing using standard APIs like (TF/Keras Op creation and convert to TFLite) is quite exorbitant, but this is the necessary price we have to pay:) . If we do not do that there will be always a possibility for gap in our CI evaluations.


If we really need this flatbuffer implementation for a comparatively easier testing approach, we should use this as a last resort.
We should not use it, until it is really difficult to reproduce the scenario using standard APIs. But I am afraid, if we start using this flatbuffer framework to write test cases, then we might not be able to maintain the sanctity.

NOTE: TFLite Schema gets updated very frequently, so maintenance cost for each version upgrade sometimes will be more, if we follow this approach.

I really appreciate all the hard work you have done! 
It is quite clearly visible. I am extremely sorry for not able to vote in favor of the change. 
But I leave to other members in TVM for their expert opinions.

Thanks again for your efforts!!!

**Comment by mbaret** [comment]
Thanks for your comments @ANSHUMAN87 and also thanks to @ekalda and @NicolaLancellotti for this PR.

It seems to me the point of contention here is how stable the TFLite flatbuffers are with respect to versioning. As I understand it, you believe they're not especially stable and really an implementation detail of the converters. However, I think the schema is now intended to be stable and versioned with a proper deprecation process in place for changes of behaviour (like in your PostProcess example) and backwards compatibility where appropriate. Therefore, if we can prove we support the operators in the schema (with these tests), we have good confidence we'll support everything the converters throw at us.

I'll ping @mshawcroft and @u99127 who can perhaps comment further on stability of the schema/flatbuffers and their suitability as a target for testing.

**Comment by u99127** [comment]
Thanks @mbaret for pulling me into this. 

My views on this topic are the following.

- The unit of interchange between Tensorflow Lite and TVM is the flat buffer. If TVM is able to consume everything represented in the flat buffer, it doesn’t matter what APIs created that as it is the ultimate of what can be represented by the higher level APIs. That produces an argument that this kind of approach can be used to reason about coverage of Tensorflow Lite operators at the flatbuffer level i.e. if TVM is able to consume and test the 150 operators that are representable in the schema, then the frontend is pretty robust in what it can consume. 

- The point about semantic changes between versions is possibly a valid one - but I would personally expect that Tensorflow lite did not break backwards compatibility (modulo actual bug fixes and fixing correctness issues with existing operators) for a standard Operator (i.e. not in the custom op space) in the way described in the example with the custom operator as that would need consumers needing updates in their tflite readers and runtimes / deployment scenarios that are independent of each other.

- Finally the notion that this somehow makes life harder for TVM developers is debatable as far as I am concerned.  I expect the complexity of understanding the semantics of the operator have changed in the implementation in the Tensor-flow lite framework to be same in both approaches. I think is easier to explore with a frozen flat buffer between 2 versions rather than having to dig in to the TF and keras API and understand the vagaries of that with differences in versions.


regards
Ramana

File: python/tvm/relay/testing/tflite.py [hunk 1]
```diff
@@ -0,0 +1,468 @@
...
+    SIGN_BIT = "SIGN_BIT"
+
+
+class Quantization:
+    "A class representing quantization of a tensor"
+
+    def __init__(
+        self,
+        scale: List[float],
+        zero_point: List[int],
```
Comment (mbaret) [inline python/tvm/relay/testing/tflite.py:51]: Given these are both lists, perhaps scales/zero_points?

File: python/tvm/relay/testing/tflite.py [hunk 2]
```diff
@@ -0,0 +1,468 @@
...
+        quantized_dimension: int = 0,
+    ):
+        """
+        Parameters
+        ----------
+        scale: List[float]
+            The scale(s)
+        zero_point: List[int]
+            The zero point(s)
```
Comment (mbaret) [inline python/tvm/relay/testing/tflite.py:60]: In what cases do we see multiple zero points?

File: python/tvm/relay/testing/tflite.py [hunk 3]
```diff
@@ -0,0 +1,468 @@
...
+
+    Returns
+    ------------
+    TensorFlow Lite model as bytes
+    """
+    tmp_dir = tempfile.gettempdir()
+
+    schema_path = tmp_dir + "/schema.fbs"
+
+    download(SCHEMA_URL, schema_path)
```
Comment (mbaret) [inline python/tvm/relay/testing/tflite.py:167]: Just to confirm, does this cache the schema between runs of this function or will it try and download it every time?

File: python/tvm/relay/testing/tflite.py [hunk 4]
```diff
@@ -0,0 +1,468 @@
...
+        "inputs": operator.op_inputs_idx,
+        "outputs": operator.op_outputs_idx,
+        "mutating_variable_inputs": [],
+    }
+    if operator.options_type != "":
+        op["builtin_options_type"] = operator.options_type
+        op["builtin_options"] = operator.options
+
+    dictionary = {
+        "version": 3,
```
Comment (mbaret) [inline python/tvm/relay/testing/tflite.py:231]: Perhaps this 3 shouldn't be hard-coded here. I'm not entirely sure what it refers to, but if it's schema versioning maybe a global at the top of the file is more appropriate.

File: python/tvm/relay/testing/tflite.py [hunk 5]
```diff
@@ -0,0 +1,468 @@
...
+    shape : List[int]
+        Shape of the tensor to be filled
+
+    Returns
+    -------
+    data_uint8.tolist() : List[int]
+        Buffer data in uint8
+    """
+    shape_multiplier = np.prod(shape)
+    data = np.random.randint(data_low, high=data_high, size=[shape_multiplier], dtype=data_type)
```
Comment (mbaret) [inline python/tvm/relay/testing/tflite.py:268]: Random data like this should probably be seeded so it's repeatable.

File: python/tvm/relay/testing/tflite.py [hunk 6]
```diff
@@ -0,0 +1,468 @@
...
+    """
+
+    try:
+        type_info = np.iinfo(dtype)
+    except ValueError:
+        type_info = np.finfo(dtype)
+    return type_info.min, type_info.max
+
+
+def get_output_qnn_params(
```
Comment (mbaret) [inline python/tvm/relay/testing/tflite.py:299]: Given this only works for convolution, consider renaming to get_conv_output_qnn_params. It's also not immediately obvious to me that this function should live here. It's applicability is quite narrow and it doesn't directly relate to constructing schema ops. Maybe being in test_forward directly might make more sense.

File: tests/python/frontend/tflite/test_forward.py [hunk 7]
```diff
@@ -868,74 +869,177 @@ def test_forward_l2_pool2d():
...
-    keras_model = tf.keras.models.Model(data_in, conv)
-    keras_model.layers[1].set_weights([kernel])
+    dtype_min, dtype_max = test_tflite.get_range_for_dtype_str(dtype)
+    channels = weights_shape[0]
 
-    # To create quantized values with dynamic range of activations, needs representative dataset
-    def representative_data_gen():
-        for i in range(1):
-            yield [data]
+    input_scale = np.random.random() * 0.1
```
Comment (mbaret) [inline tests/python/frontend/tflite/test_forward.py:886]: Setting a seed is good practice here

File: tests/python/frontend/tflite/test_forward.py [hunk 8]
```diff
@@ -868,74 +869,177 @@ def test_forward_l2_pool2d():
...
 
-    tflite_model_quant = _quantize_keras_model(keras_model, representative_data_gen)
+    # Weights in TFLite 2 are symmetric, i.e the zero point is at 0
+    if quantize_per_channel:
+        weights_scale = [np.random.random() * 0.1 for i in range(channels)]
+        weights_zp = [0 for i in range(channels)]
+    else:
+        weights_scale = [np.random.random() * 0.1]
+        weights_zp = [0]
+    weights_quantization = test_tflite.Quantization(
```
Comment (mbaret) [inline tests/python/frontend/tflite/test_forward.py:901]: In general I'm a bit concerned by how verbose you need to be to create a single operator - especially vs. the code we're replacing here. Can we think of any ways to reduce the amount of boilerplate required here to create the Tensors?

File: python/tvm/relay/testing/tflite.py (hunk 1 above)
Comment (ekalda) [reply python/tvm/relay/testing/tflite.py:51]: Yeah could do. I was thinking though that since we have multiple scales and zero points only when we have per channel quantization (i.e. only for the convolutions), it would not make much sense for all the other operators. I was also thinking to change it to accept both, a scalar and a list and convert the scalar to a list inside the class.

File: python/tvm/relay/testing/tflite.py (hunk 2 above)
Comment (ekalda) [reply python/tvm/relay/testing/tflite.py:60]: In case of per channel quantization where we have several scales, the list of zero points needs to be the same length as the list of scales. But that will just be a list of zeros, so technically we don't have multiple zero points. I'm thinking of changing it such that you can pass just one scalar zero point and if the there are more than one scales, broadcast that zero point to a list of a correct size.

File: python/tvm/relay/testing/tflite.py (hunk 3 above)
Comment (ekalda) [reply python/tvm/relay/testing/tflite.py:167]: I will cache it! I'll add a comment in the code

File: python/tvm/relay/testing/tflite.py (hunk 4 above)
Comment (ekalda) [reply python/tvm/relay/testing/tflite.py:231]: Yeah it's a schema versioning, I'll add it as a global variable to the top.

File: python/tvm/relay/testing/tflite.py (hunk 5 above)
Comment (ekalda) [reply python/tvm/relay/testing/tflite.py:268]: Will do!

File: python/tvm/relay/testing/tflite.py (hunk 6 above)
Comment (ekalda) [reply python/tvm/relay/testing/tflite.py:299]: Yeah I agree that it should be renamed to reflect its conv-ness. It would probably make sense to move it to test_forward indeed, we might also want to think if we can maybe make it shorter, e.g by specializing it to int8 (then we always know the zero point and dtype limits) or if we make the test deterministic, we could get rid of it altogether :) 

Another option I can think of is that since it is convolution specific, we can attach it as a function to the operator classes and maybe have it return an output tensor with the legit QNN params.

Any thoughts?

File: tests/python/frontend/tflite/test_forward.py (hunk 7 above)
Comment (ekalda) [reply tests/python/frontend/tflite/test_forward.py:886]: Will do!

File: tests/python/frontend/tflite/test_forward.py (hunk 8 above)
Comment (ekalda) [reply tests/python/frontend/tflite/test_forward.py:901]: Some ways we could reduce the amount of code that I can come up with:
(1) Use the same function for both, conv2d and depthwise2d and have a parameter to distinguish between the two cases, to make sure we set the quantized dimension etc correctly. 
(2) Since TFLite2 is focusing on int8, we could make it a default data type for tensors, in most cases that would mean that there will be one parameter less to set when creating a tensor.
(3) We can give the operator classes the power to create tensors themselves that make sense for the specific operator, e.g.
conv2d = test_tflite.Conv2DOperator(...)
weights_tensor = conv2d.make_weights(shape, <some optional arguments>)
We can have default data type, qnn params that make sense and optionally fill it with data. Going further, if we tag a tensor with whether it is an input or output, we could save tensors into the op object itself and extract them during json creation. Then we'd essentially only need to pass conv2d to the generate_tflite_model. We might want to think whether it is a good idea to introduce that kind of coupling though. 

In general, I'm note sure you can create a tensor directly without specifying its qnn params, shape and dtype, but maybe you have some more ideas?

**Comment by u99127** [comment]
this looks like it could do with another CI kicking after any updates you want to make @ekalda

File: tests/python/frontend/tflite/test_forward.py (hunk 8 above)
Comment (mbaret) [reply tests/python/frontend/tflite/test_forward.py:901]: I like the idea of the operators being able to create their own weights/biases. That is somewhat analogous to what happens when you create the operators with Keras/TF. This way if we need to create other Conv2D tests, we don't need to replicate a lot of boilerplate. I'm also inclined to agree that we can default the data type to int8.

File: python/tvm/relay/testing/tflite.py (hunk 6 above)
Comment (mbaret) [reply python/tvm/relay/testing/tflite.py:299]: If we move tensor creation into the operator, we could move this in too like you suggest.

**Comment by ekalda** [comment]
Closing this since our testing strategy has changed, so this change is no longer necessary.
